packages/react-components/react-card/src/components/CardHeader/useCardHeader.ts

```diff
@@ -62,14 +62,22 @@ export const useCardHeader_unstable = (props: CardHeaderProps, ref: React.Ref<HT
   const hasChildId = React.useRef(false);
   const generatedId = useId(cardHeaderClassNames.header, referenceId);
 
+  const headerSlot = resolveShorthand(header, {
+    required: true,
+    defaultProps: {
+      ref: headerRef,
+      id: !hasChildId.current ? referenceId : undefined,
+    },
+  });
+
   React.useEffect(() => {
     const headerId = !hasChildId.current ? headerRef.current?.id : undefined;
-    const childWithId = getChildWithId(header);
+    const childWithId = getChildWithId(headerSlot?.children);
 
     hasChildId.current = Boolean(childWithId);
 
     setReferenceId(getReferenceId(headerId, childWithId, generatedId));
-  }, [generatedId, header, setReferenceId]);
+  }, [generatedId, header, headerSlot, setReferenceId]);
```

Comment:
`headerSlot` will be a new object on every render, so using it as a dependency on the effect will cause the effect to run every render.

One thing that should be stable between renders is `childWithId?.props.id`. I don't think(?) you need to wait for the effect to run in order to get that value. You could do something like:

```ts
  const headerSlot = resolveShorthand(header, {
    required: true,
    defaultProps: {
      ref: headerRef,
      id: !hasChildId.current ? referenceId : undefined,
    },
  });

  const headerChildId = getChildWithId(headerSlot?.children)?.props.id;
  hasChildId.current = Boolean(headerChildId);

  React.useEffect(() => {
    const headerId = !hasChildId.current ? headerRef.current?.id : undefined;
    setReferenceId(getReferenceId(headerId, headerChildId, generatedId));
  }, [generatedId, headerChildId, setReferenceId]);
```

Reply:
And looking through it even further, I'm not sure the effect is necessary? `headerRef.current.id` should be the same as `headerSlot.id`. So this all boils down to: (a) set a generated ID on headerSlot if it doesn't have one, and (b) call setReferenceId with either the child's ID if available, or the header's ID if not.

```ts
  // [delete headerRef, hasChildId, and generatedId]

  const headerSlot = resolveShorthand(header, {
    required: true,
    defaultProps: {
      id: useId(cardHeaderClassNames.header),
    },
  });

  const headerChildId = getChildWithId(headerSlot?.children)?.props.id;

  setReferenceId(headerChildId || headerSlot.id);
```

Reply:
That makes sense. I'll look at a way to improve it by refactoring this component.